File: app/sites/imgur.py
from enum import Enum
from aiohttp import ClientSession
from collections import namedtuple
from typing import Any
from urllib.parse import urlparse
import os.path

from app.utils.download import download_binary
from app.utils.path import filename_normalize, mkdir
from app.utils.print import print_inline_end
import app.cache as cache

SLUG = 'imgur'
API_URL = 'https://api.imgur.com/3/{type}/{id}'
# client_id just from devtools
HEADERS = {
	'authorization': 'Client-ID 546c25a59c58ad7'
}

Parsed = namedtuple('Parsed', ['id', 'type'], defaults=[None, None])

class LinkType(str, Enum):
	album = 'album'
	image = 'image'

def parse_link(url: str):
	parsed = urlparse(url)
	path = parsed.path.lstrip('/').split('/')

	if path[0] == 'a' or path[0] == 'gallery':
		# https://imgur.com/a/<id>
		# https://imgur.com/gallery/<id>
		return Parsed(path[1], LinkType.album)

	if path[0] == 't':
		# https://imgur.com/t/<tag>/<id>
		return Parsed(path[2], LinkType.album)

	if len(path) == 1:
		# https://imgur.com/<id>
		return Parsed(path[0], LinkType.image)

	return Parsed()

async def fetch_info(session: ClientSession, album: Parsed) -> Any:
	url = API_URL.format(id=album.id, type=album.type)
	async with session.get(url, headers=HEADERS) as response:
		response.raise_for_status()
		info = (await response.json())['data']

	if info.get('is_album', False) is False:
		# the same model, so just put to array
		info['images'] = [info]

	return {
		'id': info['id'],
		'title': info['title'] or '',
		'images': list({
			'id': image['id'],
			'link': image['link'],
			'ext': os.path.splitext(image['link'])[-1],
			'title': image['title'] or '',
		} for image in info['images']),
	}

async def download_art(
	session: ClientSession,
	link: str,
	save_folder: str,
	name: str,
	indent=False
):
	indent_str = '  ' if indent else ''
	filename = os.path.join(save_folder, name)
	if os.path.exists(filename):
		return print(indent_str + 'Skip existing:', name)

	print_inline_end(indent_str + 'Download', name)
	await download_binary(session, link, filename)
	print('OK')

async def download(urls: list[str], data_folder: str):
	sep = ' - '

	async with ClientSession() as session:
		for url in urls:
			parsed = parse_link(url)

			if parsed.id is None:
				print('Unsupported link', url)
				continue

			cached = cache.select(SLUG, parsed.id, as_json=True)

			if cached is None:
				info = await fetch_info(session, parsed)
				cache.insert(SLUG, parsed.id, info, as_json=True)
			else:
				info = cached

			images = info['images']
			one_image = len(images) == 1
			title_prefix = sep.join([info['title'], info['id']]).strip(sep)
			title_prefix = filename_normalize(title_prefix)

			if one_image:
				save_folder = data_folder
			else:
				print(title_prefix)
				# save to sub-folder
				save_folder = os.path.join(data_folder, title_prefix)
				title_prefix = ''
			mkdir(save_folder)

			for image in images:
				title = (
					sep
					.join([title_prefix, image['title'], image['id']])
					.strip(sep)
					.replace(sep * 2, sep)
				)
				await download_art(
					session,
					image['link'],
					save_folder,
					title + image['ext'],
					indent=not one_image
				)

File: src/extractor/imgur.rs
use std::{fmt::Display, path::PathBuf};

use reqwest::{header::AUTHORIZATION, Client};
use serde::Deserialize;
use url::Url;

use super::common::{ExtractedInfo, ExtractedItem, Extractor, ExtractorOptions, ExtractorSlug};
use crate::utils::template::Template;
use crate::{CrateError, Result};

const API_ALBUM_URL: &str = "https://api.imgur.com/3/{type}/{id}";
// just from devtools
const AUTHORIZATION_KEY: &str = "Client-ID 546c25a59c58ad7";

const SLUG: ExtractorSlug = ExtractorSlug::Imgur;
const SAVE_SINGLE_FILE_PATTERN: &str =
    "{album_title}{sep}{album_id}{sep}{image_title}{sep}{image_id}.{ext}";
const SAVE_ALBUM_PATTERN: &str = "{album_title}{sep}{album_id}/{image_title}{sep}{image_id}.{ext}";

#[derive(Debug)]
pub struct ImgurExtractor {}

#[derive(Debug, Deserialize)]
struct ImgurInfoWrapper {
    data: ImgurInfoData,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum ImgurInfoData {
    Multiple(ImgurInfo),
    Single(ImgurInfoImage),
}

#[derive(Debug, Default, Deserialize)]
struct ImgurInfo {
    id: String,
    title: Option<String>,
    images: Vec<ImgurInfoImage>,
}

#[derive(Debug, Deserialize)]
struct ImgurInfoImage {
    id: String,
    link: String,
    #[serde(rename(deserialize = "type"))]
    content_type: String,
    title: Option<String>,
}

impl From<ImgurInfoImage> for ImgurInfo {
    /// Put single image to array, because it the same model
    fn from(value: ImgurInfoImage) -> Self {
        Self {
            id: value.id.clone(),
            title: value.title.clone(),
            images: vec![value],
        }
    }
}

#[derive(Debug)]
enum LinkType {
    Album,
    Image,
}

impl Display for LinkType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match &self {
            Self::Album => "album",
            Self::Image => "image",
        };
        write!(f, "{s}")
    }
}

#[derive(Debug)]
struct Parsed {
    id: String,
    link_type: LinkType,
}

impl Parsed {
    fn new(id: &str, link_type: LinkType) -> Self {
        let id = id.into();
        Self { id, link_type }
    }
}

impl TryFrom<&str> for Parsed {
    type Error = CrateError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        let parsed = Url::parse(value).map_err(|s| CrateError::invalid_url(s.to_string()))?;
        let Some(segments) = parsed.path_segments().map(|c| c.collect::<Vec<_>>()) else {
            let msg = format!("cannot get path segments from {value}");
            return Err(CrateError::plain(msg));
        };

        match segments.as_slice() {
            ["" | "a" | "gallery" | "t"] | ["t", _] => Err(CrateError::invalid_url(value)),

            // https://imgur.com/<id>
            [id] => Ok(Parsed::new(id, LinkType::Image)),

            // https://imgur.com/a/<id>
            // https://imgur.com/gallery/<id>
            // https://imgur.com/t/<tag>/<id>
            ["a" | "gallery", id] | ["t", _, id] => Ok(Parsed::new(id, LinkType::Album)),

            _ => Err(CrateError::unsupported_url(value)),
        }
    }
}

impl Extractor for ImgurExtractor {
    async fn fetch_info(
        urls: &[&str],
        config: &ExtractorOptions,
    ) -> Result<Vec<impl ExtractedInfo>> {
        let url_template = Template::try_from(API_ALBUM_URL)?;
        let client = Client::new();

        let mut extracted = vec![];
        for &url in urls {
            let parsed = match Parsed::try_from(url) {
                Ok(p) => p,
                Err(e) => {
                    eprintln!("cannot parse url: {e}");
                    continue;
                }
            };
            let sub = [("id", parsed.id), ("type", parsed.link_type.to_string())];
            let url = url_template.render(sub)?;
            match Self::fetch_info_inner(&client, &url).await {
                Ok(info) => extracted.push(info),
                Err(e) => eprintln!("cannot fetch info: {e}"),
            }
        }
        Ok(extracted)
    }
}

impl ImgurExtractor {
    async fn fetch_info_inner(client: &Client, url: &str) -> Result<ImgurInfo> {
        let res = client
            .get(url)
            .header(AUTHORIZATION, AUTHORIZATION_KEY)
            .send()
            .await
            .map_err(CrateError::ReqwestError)?;
        let res: ImgurInfoWrapper = res.json().await.map_err(CrateError::ReqwestError)?;
        let data = match res.data {
            ImgurInfoData::Single(image) => ImgurInfo::from(image),
            ImgurInfoData::Multiple(info) => info,
        };
        Ok(data)
    }
}

impl ExtractedInfo for ImgurInfo {}

impl IntoIterator for ImgurInfo {
    type Item = ExtractedItem;
    type IntoIter = ImgurInfoIter;

    fn into_iter(self) -> Self::IntoIter {
        let template = if self.images.len() == 1 {
            SAVE_SINGLE_FILE_PATTERN
        } else {
            SAVE_ALBUM_PATTERN
        };
        let template = Template::try_from(template).unwrap();
        Self::IntoIter {
            data: self,
            template,
            at: 0,
        }
    }
}

#[derive(Debug)]
struct ImgurInfoIter {
    data: ImgurInfo,
    template: Template,
    at: usize,
}

impl Iterator for ImgurInfoIter {
    type Item = ExtractedItem;

    fn next(&mut self) -> Option<Self::Item> {
        let Some(image) = self.data.images.get(self.at) else {
            return None;
        };

        let folder = PathBuf::from(SLUG.to_string());
        let ext = image.content_type.split('/').last().unwrap_or_default();
        let sub: [(&str, &str); 6] = [
            ("album_title", &self.data.title.clone().unwrap_or("".into())),
            ("album_id", &self.data.id),
            ("image_title", &image.title.clone().unwrap_or("".into())),
            ("image_id", &image.id),
            ("sep", " - "),
            ("ext", ext),
        ];

        self.at += 1;
        let f = self.template.render(sub).unwrap();
        let f = f.replace(" -  - ", " - ");
        Some(Self::Item::new(&image.link, folder.join(f)))
    }
}
